db/range_del_aggregator.cc

```diff
@@ -9,233 +9,146 @@
-          icmp_.user_comparator()->Compare(parsed.user_key,
-                                           tombstone_map_iter->first) < 0) {
-        // before start of deletion intervals
-        return false;
+// An UncollapsedRangeDelMap is quick to create but slow to answer ShouldDelete
+// queries.
+class UncollapsedRangeDelMap : public RangeDelMap {
+ public:
+  UncollapsedRangeDelMap(const Comparator* ucmp)
+      : rep_(TombstoneStartKeyComparator(ucmp)), ucmp_(ucmp) {}
+
+  bool ShouldDelete(const ParsedInternalKey& parsed,
+                    RangeDelPositioningMode mode) {
+    assert(mode == RangeDelPositioningMode::kFullScan);
+    for (const auto& tombstone : rep_) {
+      if (ucmp_->Compare(parsed.user_key, tombstone.start_key_) < 0) {
+        break;
       }
-      break;
-    case RangeDelPositioningMode::kBinarySearch:
-      assert(collapse_deletions_);
-      tombstone_map_iter =
-          tombstone_map.upper_bound(parsed.user_key);
-      if (tombstone_map_iter == tombstone_map.begin()) {
-        // before start of deletion intervals
-        return false;
+      if (parsed.sequence < tombstone.seq_ &&
+          ucmp_->Compare(parsed.user_key, tombstone.end_key_) < 0) {
+        return true;
       }
-      --tombstone_map_iter;
-      break;
-  }
-  assert(mode != RangeDelPositioningMode::kFullScan);
-  assert(tombstone_map_iter != tombstone_map.end() &&
-         icmp_.user_comparator()->Compare(tombstone_map_iter->first,
-                                          parsed.user_key) <= 0);
-  assert(std::next(tombstone_map_iter) == tombstone_map.end() ||
-         icmp_.user_comparator()->Compare(
-             parsed.user_key, std::next(tombstone_map_iter)->first) < 0);
-  return parsed.sequence < tombstone_map_iter->second.seq_;
-}
-
-bool RangeDelAggregator::IsRangeOverlapped(const Slice& start,
-                                           const Slice& end) {
-  // so far only implemented for non-collapsed mode since file ingestion (only
-  //  client) doesn't use collapsing
-  assert(!collapse_deletions_);
-  if (rep_ == nullptr) {
+    }
     return false;
   }
-  for (const auto& seqnum_and_tombstone_map : rep_->stripe_map_) {
-    for (const auto& start_key_and_tombstone :
-         seqnum_and_tombstone_map.second.raw_map) {
-      const auto& tombstone = start_key_and_tombstone.second;
-      if (icmp_.user_comparator()->Compare(start, tombstone.end_key_) < 0 &&
-          icmp_.user_comparator()->Compare(tombstone.start_key_, end) <= 0 &&
-          icmp_.user_comparator()->Compare(tombstone.start_key_,
-                                           tombstone.end_key_) < 0) {
+
+  bool IsRangeOverlapped(const Slice& start, const Slice& end) {
+    for (const auto& tombstone : rep_) {
+      if (ucmp_->Compare(start, tombstone.end_key_) < 0 &&
+          ucmp_->Compare(tombstone.start_key_, end) <= 0 &&
+          ucmp_->Compare(tombstone.start_key_, tombstone.end_key_) < 0) {
```

Comment:
When is `tombstone.start_key_ >= tombstone.end_key_`? I can see how a user might accidentally create such tombstones, but why are they ever persisted?